File: python/task1/S1_VAR_08.py
# S1_VAR_08 - Big integers (no MAX_SAFE_INTEGER problem)

big_int = 10**100
print(f"10**100 = {big_int}")
print(f"Type: {type(big_int).__name__}")
print(f"Number of digits: {len(str(big_int))}")
print()

# Compare with float
big_float = float(10**100)
print(f"float(10**100) = {big_float}")
print(f"Type: {type(big_float).__name__}")

# Precision comparison
print()
print("Precision issue with floats:")
print(f"10**100 + 1 (int): {10**100 + 1}")
print(f"float(10**100) + 1: {float(10**100) + 1}")
# Floats lose precision for very large numbers

File: js/task1/S1_VAR_08.js
// S1_VAR_08 - Big integers
// Numbers beyond MAX_SAFE_INTEGER lose precision, use BigInt for exact math

function safeAdd(a, b) {
  const needsBigInt =
    Math.abs(a) >= Number.MAX_SAFE_INTEGER ||
    Math.abs(b) >= Number.MAX_SAFE_INTEGER;

  if (needsBigInt) {
    const result = BigInt(a) + BigInt(b);
    console.log(`  -> Used BigInt (returned type: bigint)`);
    return result;
  }

  console.log(`  -> Used Number (returned type: number)`);
  return a + b;
}

// Test cases
console.log("Testing safeAdd():\n");

console.log("safeAdd(10, 20):");
console.log("  =", safeAdd(10, 20));

console.log("\nsafeAdd(Number.MAX_SAFE_INTEGER, 1):");
console.log("  =", safeAdd(Number.MAX_SAFE_INTEGER, 1));

console.log("\nsafeAdd(9007199254740991, 9007199254740991):");
console.log("  =", safeAdd(9007199254740991, 9007199254740991));

// Show why this matters
console.log("\n--- Why BigInt matters ---");
console.log("MAX_SAFE_INTEGER:", Number.MAX_SAFE_INTEGER);
console.log("MAX_SAFE_INTEGER + 1 (Number):", Number.MAX_SAFE_INTEGER + 1);
console.log("MAX_SAFE_INTEGER + 2 (Number):", Number.MAX_SAFE_INTEGER + 2);
console.log("Notice: +1 and +2 give the same result (precision loss)");
